File: vtnext/client-py/vtnext/commands.py
"""
VTNext output commands - escape sequences to send to the terminal.
"""

import base64
from typing import Optional, Tuple, List

ESC = "\x1b"
BEL = "\x07"
PREFIX = f"{ESC}]vtn;"


def text(
    content: str,
    x: float = 0,
    y: float = 0,
    z: float = 0,
    rotation: float = 0,
    scale: float = 1.0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
) -> str:
    """Draw text at arbitrary position with optional rotation and scale."""
    r, g, b, a = color
    safe_content = content.replace('"', '\\"')
    return f'{PREFIX}text;{x};{y};{z};{rotation};{scale};{r};{g};{b};{a};"{safe_content}"{BEL}'


def char(
    codepoint: int,
    x: float = 0,
    y: float = 0,
    z: float = 0,
    rotation: float = 0,
    scale_x: float = 1.0,
    scale_y: float = 1.0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
) -> str:
    """Draw a single character (by codepoint) with full transform control."""
    r, g, b, a = color
    return f"{PREFIX}char;{x};{y};{z};{rotation};{scale_x};{scale_y};{r};{g};{b};{a};{codepoint}{BEL}"


def rect(
    x: float,
    y: float,
    width: float,
    height: float,
    z: float = 0,
    rotation: float = 0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
    filled: bool = True,
) -> str:
    """Draw a rectangle."""
    r, g, b, a = color
    f = 1 if filled else 0
    return f"{PREFIX}rect;{x};{y};{z};{width};{height};{rotation};{r};{g};{b};{a};{f}{BEL}"


def line(
    x1: float,
    y1: float,
    x2: float,
    y2: float,
    z: float = 0,
    thickness: float = 1.0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
) -> str:
    """Draw a line."""
    r, g, b, a = color
    return f"{PREFIX}line;{x1};{y1};{x2};{y2};{z};{thickness};{r};{g};{b};{a}{BEL}"


def circle(
    cx: float,
    cy: float,
    radius: float,
    z: float = 0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
    filled: bool = True,
) -> str:
    """Draw a circle."""
    r, g, b, a = color
    f = 1 if filled else 0
    return f"{PREFIX}circle;{cx};{cy};{z};{radius};{r};{g};{b};{a};{f}{BEL}"


def ellipse(
    cx: float,
    cy: float,
    rx: float,
    ry: float,
    z: float = 0,
    rotation: float = 0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
    filled: bool = True,
) -> str:
    """Draw an ellipse."""
    r, g, b, a = color
    f = 1 if filled else 0
    return f"{PREFIX}ellipse;{cx};{cy};{z};{rx};{ry};{rotation};{r};{g};{b};{a};{f}{BEL}"


def arc(
    cx: float,
    cy: float,
    radius: float,
    start_angle: float,
    end_angle: float,
    z: float = 0,
    thickness: float = 2.0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
) -> str:
    """Draw an arc."""
    r, g, b, a = color
    return f"{PREFIX}arc;{cx};{cy};{z};{radius};{start_angle};{end_angle};{thickness};{r};{g};{b};{a}{BEL}"


def polygon(
    points: List[Tuple[float, float]],
    z: float = 0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
    filled: bool = True,
) -> str:
    """Draw a polygon from a list of (x, y) points."""
    r, g, b, a = color
    f = 1 if filled else 0
    n = len(points)
    pts_str = ";".join(f"{p[0]};{p[1]}" for p in points)
    return f"{PREFIX}poly;{z};{r};{g};{b};{a};{f};{n};{pts_str}{BEL}"


def rounded_rect(
    x: float,
    y: float,
    width: float,
    height: float,
    radius: float,
    z: float = 0,
    color: Tuple[int, int, int, int] = (255, 255, 255, 255),
    filled: bool = True,
) -> str:
    """Draw a rounded rectangle."""
    r, g, b, a = color
    f = 1 if filled else 0
    return f"{PREFIX}rrect;{x};{y};{z};{width};{height};{radius};{r};{g};{b};{a};{f}{BEL}"


def clear(
    color: Tuple[int, int, int, int] = (0, 0, 0, 255),
    region: Optional[Tuple[float, float, float, float]] = None,
) -> str:
    """Clear the viewport or a region."""
    r, g, b, a = color
    if region:
        rx, ry, rw, rh = region
        return f"{PREFIX}clear;{rx};{ry};{rw};{rh};{r};{g};{b};{a}{BEL}"
    return f"{PREFIX}clear;{r};{g};{b};{a}{BEL}"


def viewport(width: float, height: float) -> str:
    """Set the logical viewport size."""
    return f"{PREFIX}viewport;{width};{height}{BEL}"


def image(
    data: bytes,
    x: float = 0,
    y: float = 0,
    z: float = 0,
    width: float = 0,
    height: float = 0,
    rotation: float = 0,
    format: str = "png",
) -> str:
    """Draw an image."""
    b64 = base64.b64encode(data).decode("ascii")
    return f"{PREFIX}img;{x};{y};{z};{width};{height};{rotation};{format};{b64}{BEL}"


def layer_push(
    layer_id: str,
    x: float = 0,
    y: float = 0,
    width: float = 0,
    height: float = 0,
    opacity: float = 1.0,
) -> str:
    """Push a new layer."""
    return f"{PREFIX}layer;push;{layer_id};{x};{y};{width};{height};{opacity}{BEL}"


def layer_pop(layer_id: str) -> str:
    """Pop a layer."""
    return f"{PREFIX}layer;pop;{layer_id}{BEL}"


def query_version() -> str:
    """Query terminal version."""
    return f"{PREFIX}query;version{BEL}"


def query_size() -> str:
    """Query viewport size."""
    return f"{PREFIX}query;size{BEL}"


def query_features() -> str:
    """Query supported features."""
    return f"{PREFIX}query;features{BEL}"


def input_raw() -> str:
    """Enable raw input mode (game-engine style)."""
    return f"{PREFIX}input;raw{BEL}"


def input_normal() -> str:
    """Enable normal input mode (traditional terminal)."""
    return f"{PREFIX}input;normal{BEL}"


def cursor_hide() -> str:
    """Hide the cursor."""
    return f"{PREFIX}cursor;hide{BEL}"


def cursor_show() -> str:
    """Show the cursor."""
    return f"{PREFIX}cursor;show{BEL}"

File: vtnext/renderer/src/protocol/commands.rs
//! VTNext output commands (terminal -> renderer)

use super::types::*;

/// All possible VTNext output commands
#[derive(Debug, Clone)]
pub enum Command {
    /// Draw text at position
    Text {
        transform: Transform,
        color: Color,
        content: String,
    },

    /// Draw single character
    Char {
        transform: Transform,
        color: Color,
        codepoint: u32,
    },

    /// Draw image
    Image {
        x: f32,
        y: f32,
        z: f32,
        width: f32,
        height: f32,
        rotation: f32,
        format: ImageFormat,
        data: Vec<u8>,
    },

    /// Draw rectangle
    Rect {
        x: f32,
        y: f32,
        z: f32,
        width: f32,
        height: f32,
        rotation: f32,
        color: Color,
        filled: bool,
    },

    /// Draw line
    Line {
        x1: f32,
        y1: f32,
        x2: f32,
        y2: f32,
        z: f32,
        thickness: f32,
        color: Color,
    },

    /// Draw circle
    Circle {
        cx: f32,
        cy: f32,
        z: f32,
        radius: f32,
        color: Color,
        filled: bool,
    },

    /// Draw ellipse
    Ellipse {
        cx: f32,
        cy: f32,
        z: f32,
        rx: f32,
        ry: f32,
        rotation: f32,
        color: Color,
        filled: bool,
    },

    /// Draw arc
    Arc {
        cx: f32,
        cy: f32,
        z: f32,
        radius: f32,
        start_angle: f32,
        end_angle: f32,
        thickness: f32,
        color: Color,
    },

    /// Draw polygon
    Polygon {
        points: Vec<(f32, f32)>,
        z: f32,
        color: Color,
        filled: bool,
    },

    /// Draw rounded rectangle
    RoundedRect {
        x: f32,
        y: f32,
        z: f32,
        width: f32,
        height: f32,
        radius: f32,
        color: Color,
        filled: bool,
    },

    /// Clear viewport
    Clear {
        color: Color,
        region: Option<(f32, f32, f32, f32)>, // x, y, w, h
    },

    /// Set viewport size
    Viewport {
        width: f32,
        height: f32,
    },

    /// Layer operations
    LayerPush {
        id: String,
        x: f32,
        y: f32,
        width: f32,
        height: f32,
        opacity: f32,
    },
    LayerPop {
        id: String,
    },

    /// Legacy VT100 subwindow
    LegacyCreate {
        id: String,
        x: f32,
        y: f32,
        cols: u32,
        rows: u32,
        font_size: f32,
    },
    LegacyDestroy {
        id: String,
    },
    LegacyWrite {
        id: String,
        data: Vec<u8>,
    },
    LegacyFocus {
        id: String,
    },

    /// Query commands
    QueryVersion,
    QuerySize,
    QueryFeatures,

    /// Input mode
    InputMode {
        raw: bool,
    },

    /// Cursor control
    CursorHide,
    CursorShow,
    CursorSet {
        image_data: Vec<u8>,
        hotspot_x: f32,
        hotspot_y: f32,
    },
}

/// Input events (renderer -> application)
#[derive(Debug, Clone)]
pub enum InputEvent {
    /// Key pressed
    KeyDown {
        keycode: u32,
        scancode: u32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Key released
    KeyUp {
        keycode: u32,
        scancode: u32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Key repeat
    KeyRepeat {
        keycode: u32,
        scancode: u32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Character input (for text)
    KeyChar {
        codepoint: u32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Mouse button pressed
    MouseDown {
        button: MouseButton,
        x: f32,
        y: f32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Mouse button released
    MouseUp {
        button: MouseButton,
        x: f32,
        y: f32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Mouse moved
    MouseMove {
        x: f32,
        y: f32,
        dx: f32,
        dy: f32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Mouse wheel/scroll
    Scroll {
        dx: f32,
        dy: f32,
        x: f32,
        y: f32,
        modifiers: Modifiers,
        timestamp: u64,
    },

    /// Window focus
    Focus {
        focused: bool,
    },

    /// Window resize
    Resize {
        width: f32,
        height: f32,
    },
}

impl InputEvent {
    /// Encode event as escape sequence to send to application
    pub fn encode(&self) -> String {
        match self {
            InputEvent::KeyDown { keycode, scancode, modifiers, timestamp } => {
                format!("\x1b]vtni;kd;{};{};{};{}\x07", keycode, scancode, modifiers.0, timestamp)
            }
            InputEvent::KeyUp { keycode, scancode, modifiers, timestamp } => {
                format!("\x1b]vtni;ku;{};{};{};{}\x07", keycode, scancode, modifiers.0, timestamp)
            }
            InputEvent::KeyRepeat { keycode, scancode, modifiers, timestamp } => {
                format!("\x1b]vtni;kr;{};{};{};{}\x07", keycode, scancode, modifiers.0, timestamp)
            }
            InputEvent::KeyChar { codepoint, modifiers, timestamp } => {
                format!("\x1b]vtni;kc;{};{};{}\x07", codepoint, modifiers.0, timestamp)
            }
            InputEvent::MouseDown { button, x, y, modifiers, timestamp } => {
                let btn = match button {
                    MouseButton::Left => 0,
                    MouseButton::Right => 1,
                    MouseButton::Middle => 2,
                    MouseButton::Extra(n) => *n,
                };
                format!("\x1b]vtni;md;{};{};{};{};{}\x07", btn, x, y, modifiers.0, timestamp)
            }
            InputEvent::MouseUp { button, x, y, modifiers, timestamp } => {
                let btn = match button {
                    MouseButton::Left => 0,
                    MouseButton::Right => 1,
                    MouseButton::Middle => 2,
                    MouseButton::Extra(n) => *n,
                };
                format!("\x1b]vtni;mu;{};{};{};{};{}\x07", btn, x, y, modifiers.0, timestamp)
            }
            InputEvent::MouseMove { x, y, dx, dy, modifiers, timestamp } => {
                format!("\x1b]vtni;mm;{};{};{};{};{};{}\x07", x, y, dx, dy, modifiers.0, timestamp)
            }
            InputEvent::Scroll { dx, dy, x, y, modifiers, timestamp } => {
                format!("\x1b]vtni;mw;{};{};{};{};{};{}\x07", dx, dy, x, y, modifiers.0, timestamp)
            }
            InputEvent::Focus { focused } => {
                format!("\x1b]vtni;focus;{}\x07", if *focused { 1 } else { 0 })
            }
            InputEvent::Resize { width, height } => {
                format!("\x1b]vtni;resize;{};{}\x07", width, height)
            }
        }
    }
}
